cmd/ossrecord/ossrecord.man
NAME
ossrecord - Open Sound System recording program.

SYNOPSIS
ossrecord [-Rhlv] [-F cntname | ? ] [ -c channels ] [ -d devname ]
          [ -f fmtname | ? ] [ -g gain ] [ -i recsource | ? ] [ -m nfiles ]
          [ -r command ] [ -s rate ] [ -t maxsecs ] filename | - ...


DESCRIPTION
The ossrecord program records audio in Microsoft RIFF (wav) format. It
will record from any input that's currently set as the recording source
by the ossxmix/ossmix mixer programs. With the -l option, you also get
a level meter that will display VU levels in a character mode.

The filename parameter is name of the (.wav) file to be produced. Output can be
sent to stdout by giving - as the file name.

OPTIONS
-s<rate> 	Select the recording rate for raw PCM audio (eg -s48000).
-c<channels> 	Select the number of channels 1=mono 2=stereo, 4, 6, 8, etc. 
-d<devname>  	Select <devname> as the device (eg -d/dev/dsp2).
-f<fmt>		Select the input sample format (eg -fS32_LE or -fMU_LAW)
-f?		Prints the list of supported format names.
-F<cnt>		Select the container format (eg WAV or AU). Default is WAV.
-F?		Prints the list of supported container formats.
-R		Open audio device in raw mode to disable virtual mixing and
		sample rate/format conversions. Can be used when recording
		from a digital source (S/PDIF input).
-v		Verbose output.
-l		Display level meters (character based).
-i<recsrc|?> 	Select the recording source or display available recording 
		sources if '?' is supplied.
		e.g. ossrecord -i? may display:
			        vol
        			line (currently selected)
        			mic
        			cd
        			aux1
        			phone
        			mono
        			video
-m<nfiles> 	Repeat the recording operation <nfiles> times. The filename
		argument must have %d (or %02d) somewhere in the file to
		guarantee unique filenames. If no %d is given then subsequent
		recordings will overwrite the previous one(s). This option is
		useful only with loopback audio devices or if the -t option
		is used.
-r<command> 	This option launches the <command> in background after 
		recording the file has completed. The name of the recorded file
		will be given as the (only) command line argument. When the -m 
		option is used the script will run in parallel while recording
		the next file. See the COMMAND SCRIPT section (below) for more
		info.
-g<gain>	Amplify recorded samples by percentage given as argument.
		100 (default) means normal signal level, 200 means double level.
		Only supported in 16 and 32 bit modes.
-t<maxsecs>	Do not record more than <maxsecs> seconds in a single recording
		operation.
-L<level>	Set the recording level to <level>.
-h		Display usage instructions.

COMMAND SCRIPT
		The -r command line argument makes it possible to execute a
		script or program after recording of the wave file is finished.
		Below is a simple scell script that does MP3 encoding using
		lame.

#!/bin/sh

WAVENAME=$1

MP3NAME=$1.mp3

lame -m s -h --preset studio $WAVENAME $MP3NAME

exit 0

Another example script for ossrecord is a simple CGI script for live MP3
streaming (from /dev/dsp).

#!/bin/sh

echo Content-Type: audio/mp3

echo

ossrecord -S -b16 -s48 - | lame -m j - -

exit 0

NOTES
The ossrecord executable is the same as the ossplay executable.
Behaviour is decided by the name used to invoke the program.

Some file formats allocate only 32 bits to record the file length,
which may prevent some programs from reading a too large file properly.
If a recording may pass the 4GB limit (a bit more then 6 hours and 10 minutes
of sound assuming [48Khz/16bit/stereo] quality), it's best to have ossrecord
use the AU or RAW containers (via -F switch) which do not have this restriction.

SEE ALSO
ossplay(1), ossmix(1), ossxmix(1)

FILES
/usr/bin/ossrecord

AUTHOR
4Front Technologies


Changelog
Change log for OSS
==================

This file contains a list of the changes made to OSS. Changes are sorted by
directory (not by time).

Global changes
--------------

070710 by hannu
Added "template" for Linux 2.4.x support. However for the time being it's not
functional.


cmd/ossrecord/ossrecord.c
/* Purpose: Simple wave file (.wav) recorder for OSS
 */
#define COPYING Copyright (C) Hannu Savolainen and Dev Mazumdar 2000-2006. All rights reserved.

#include <fcntl.h>
#include <unistd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <soundcard.h>
#include "ossrecord.h"
#include "ossrecord_encode.h"

int audio_fd, channels = -1, datalimit = -1, format = -1, speed = -1,
    type = WAVE_FILE;
char dspdev[512] = "/dev/dsp"; /* Must be at least OSS_DEVNODE_SIZE */
char script[512] = "";
int amplification = 100, level_meters = 0, nfiles = 1, raw_mode = 0,
    reclevel = 0, verbose = 0;
static const char * recsrc = NULL;
static const format_t format_a[] = {
  {"S8",		AFMT_S8},
  {"U8",		AFMT_U8},
  {"S16_LE",		AFMT_S16_LE},
  {"S16_BE",		AFMT_S16_BE},
  {"S24_LE",		AFMT_S24_LE},
  {"S24_BE",		AFMT_S24_BE},
  {"S32_LE",		AFMT_S32_LE},
  {"S32_BE",		AFMT_S32_BE},
  {"MU_LAW",		AFMT_MU_LAW},
  {"S24_PACKED",	AFMT_S24_PACKED},
  {NULL,		0}
};

static const container_t container_a[] = {
  {"WAV",		WAVE_FILE,	AFMT_S16_LE,	2,	48000},
  {"AU",		AU_FILE,	AFMT_S8,	1,	8000},
  {"RAW",		RAW_FILE,	AFMT_S16_LE,	2,	44100},
  {NULL,		0}
};

static void fatal_signal (int);
static void find_devname (char *, const char *);
static void intr (int);
static int select_container (const char *);
static int select_format (const char *);
static void select_recsrc (const char *);
static void usage (const char *);

static int
select_container (const char * optstr)
{
/*
 * Handling of the -F command line option (force container format).
 *
 * Empty or "?" shows the supported container format names.
 */
  int i;

  if ((!strcmp(optstr, "?")) || (*optstr == '\0'))
    {
      fprintf (stdout, "\nSupported format names are:\n\n");
      for (i = 0; container_a[i].name != NULL; i++)
        fprintf (stdout, "%s ", container_a[i].name);
      fprintf (stdout, "\n");
      exit (0);
    }

  for (i = 0; container_a[i].name != NULL; i++)
    if (!strcasecmp(container_a[i].name, optstr))
      return container_a[i].type;

  fprintf (stderr, "Unsupported container format name '%s'!\n", optstr);
  exit (-1);
}

static int
select_format (const char * optstr)
{
/*
 * Handling of the -f command line option (force output format).
 *
 * Empty or "?" shows the supported format names.
 */
  int i;

  if ((!strcmp(optstr, "?")) || (*optstr == '\0'))
    {
      fprintf (stdout, "\nSupported format names are:\n\n");
      for (i = 0; format_a[i].name != NULL; i++)
        fprintf (stdout, "%s ", format_a[i].name);
      fprintf (stdout, "\n");
      exit (0);
    }

  for (i = 0; format_a[i].name != NULL; i++)
    if (!strcasecmp(format_a[i].name, optstr))
      return format_a[i].fmt;

  fprintf (stderr, "Unsupported format name '%s'!\n", optstr);
  exit (-1);
}

static void
select_recsrc (const char * srcname)
{
/*
 * Handling of the -i command line option (recording source selection).
 *
 * Empty or "?" shows the available recording sources.
 */
  int i, src;
  oss_mixer_enuminfo ei;

  if (ioctl (audio_fd, SNDCTL_DSP_GET_RECSRC_NAMES, &ei) == -1)
    {
      perror ("SNDCTL_DSP_GET_RECSRC_NAMES");
      exit (-1);
    }

  if (ioctl (audio_fd, SNDCTL_DSP_GET_RECSRC, &src) == -1)
    {
      perror ("SNDCTL_DSP_GET_RECSRC");
      exit (-1);
    }

  if (*srcname == 0 || strcmp (srcname, "?") == 0)
    {
      fprintf (stderr, "\nPossible recording sources for the selected device:\n\n");

      for (i = 0; i < ei.nvalues; i++)
	{
	  fprintf (stderr, "\t%s", ei.strings + ei.strindex[i]);
	  if (i == src)
	    fprintf (stderr, " (currently selected)");
	  fprintf (stderr, "\n");
	}
      fprintf (stderr, "\n");
      exit (0);
    }

  for (i = 0; i < ei.nvalues; i++)
    {
      char *s = ei.strings + ei.strindex[i];
      if (strcmp (s, srcname) == 0)
	{
	  src = i;
	  if (ioctl (audio_fd, SNDCTL_DSP_SET_RECSRC, &src) == -1)
	    {
	      perror ("SNDCTL_DSP_SET_RECSRC");
	      exit (-1);
	    }
	  return;
	}
    }

  fprintf (stderr,
	   "Unknown recording source name '%s' - use -i? to get the list\n",
	   srcname);
  exit (-1);
}

void
open_audio (void)
{
  int flags = O_RDONLY;

  if (raw_mode)
    flags |= O_EXCL;

  audio_fd = open (dspdev, flags);
  if (audio_fd < 0)
    {
      perror (dspdev);
      exit (-1);
    }

  if (verbose)
     {
  	oss_audioinfo ai;

	ai.dev=-1;

	if (ioctl(audio_fd, SNDCTL_ENGINEINFO, &ai) != -1)
	   fprintf (stderr, "Recording from %s\n", ai.name);
     }

  if (raw_mode)
    {
      /*
       * Disable format conversions. No error checking.
       */
      int tmp = 0;
      ioctl (audio_fd, SNDCTL_DSP_COOKEDMODE, &tmp);
    }

  if (recsrc != NULL)
    select_recsrc (recsrc);
}

static void
intr (int i)
{
  if (verbose)
    {
      decorate (NULL, 0, 0);
      fprintf (stderr, "\nStopped (%d).\n", i);
    }
  fprintf (stderr, "\n");
  fflush (stderr);
  end ();
  if (*script)
    {
      fprintf (stderr,
               "Waiting for the '%s' script(s) to finish - please stand by\n",
               script);
      while (wait (NULL) != -1);
    }
  exit (0);
}

static void
fatal_signal (int i)
{
  if (verbose)
    {
      decorate (NULL, 0, 0);
      fprintf (stderr, "\nStopped.(signal %d)\n", i);
    }
  end ();
  if (*script)
    {
      fprintf (stderr,
               "Waiting for the '%s' script(s) to finish - please stand by\n",
               script);
      while (wait (NULL) != -1);
    }
  exit (0);
}

static void
find_devname (char * devname, const char * num)
{
/*
 * OSS 4.0 the audio device numbering may be different from the
 * legacy /dev/dsp# numbering reported by /dev/sndstat. Try to find the
 * device name (devnode) that matches the given device number.
 *
 * Prior versions of ossrecord simply used the the /dev/dsp# number.
 */
  int dev;
  int mixer_fd;
  oss_audioinfo ai;

  if (sscanf (num, "%d", &dev) != 1)
    {
      fprintf (stderr, "Invalid audio device number '%s'\n", num);
      exit (-1);
    }

  if ((mixer_fd = open ("/dev/mixer", O_RDWR, 0)) == -1)
    {
      perror ("/dev/mixer");
      fprintf (stderr, "Warning: Defaulting to /dev/dsp%s\n", num);
      sprintf (devname, "/dev/dsp%s", num);
      return;
    }

  ai.dev = dev;

  if (ioctl (mixer_fd, SNDCTL_AUDIOINFO, &ai) == -1)
    {
      perror ("/dev/mixer SNDCTL_AUDIOINFO");
      fprintf (stderr, "Warning: Defaulting to /dev/dsp%s\n", num);
      sprintf (devname, "/dev/dsp%s", num);
      close (mixer_fd);
      return;
    }

  strcpy (devname, ai.devnode);

  close (mixer_fd);
}

static void
usage (const char * prog)
{
  fprintf (stdout, "Usage: %s [options] filename\n", prog?prog:"ossrecord");
  fprintf (stdout, "  Options:  -v             Verbose output.\n");
  fprintf (stdout, "            -d<device>     Change input device.\n");
  fprintf (stdout, "            -c<channels>   Change number of channels\n");
  fprintf (stdout, "            -L<level>      Change recording level.\n");
  fprintf (stdout,
           "            -g<gain>       Change gain percentage.\n");
  fprintf (stdout, "            -s<rate>       Change recording rate.\n");
  fprintf (stdout, "            -f<fmt|?>      Change/Query sample format.\n");
  fprintf (stdout,
           "            -F<cnt|?>      Change/Query container format.\n");
  fprintf (stdout, "            -l             Display level meters.\n");
  fprintf (stdout,
           "            -i<recsrc|?>   Select/Query recording source.\n");
  fprintf (stdout,
           "            -m<nfiles>     Repeat recording <nfiles> times.\n");
  fprintf (stdout,
           "            -r<command>    Run <command> after recording.\n");
  fprintf (stdout,
           "            -t<maxsecs>    Record no more than <maxsecs> in a"
           " single recording.\n");
  exit (-1);
}

int
main (int argc, char * argv[])
{
  int c;
  extern char * optarg;
  extern int optind;

  if (argc < 2)
    usage (argv[0]);

  while ((c = getopt (argc, argv, "F:L:MRSb:c:d:f:g:hi:lm:r:s:t:wv")) != EOF)
    switch (c)
      {
        case 'F':
          type = select_container (optarg);
          break;

        case 'L':
          reclevel = atoi (optarg);
          if (reclevel < 1 || reclevel > 100)
            {
              fprintf (stderr, "%s: Bad recording level '%s'\n", argv[0],
                       optarg);
              exit (-1);
            }
          break;

        case 'M':
          channels = 1;
          break;

        case 'R':
          raw_mode = 1;
          break;

        case 'S':
          channels = 2;
          break;

        case 'b':
          c = atoi (optarg);
          c += c % 8;
          switch (c)
            {
              case 8: format = AFMT_U8; break;
              case 16: format = AFMT_S16_LE; break;
              case 24: format = AFMT_S24_LE; break;
              case 32: format = AFMT_S32_LE; break;
              default:
                fprintf (stderr, "Error: Unsupported number of bits %d\n", c);
                exit (-1);
            }
          break;

        case 'c':
          channels = atoi (optarg);
          break;

        case 'd':
          if (*optarg >= '0' && *optarg <= '9')
            /* Just the device number was given */
            find_devname (dspdev, optarg);
          else
            snprintf (dspdev, sizeof(dspdev), "%s", optarg);
          break;

        case 'f':
          format = select_format (optarg);
          break;

        case 'g':
          amplification = atoi (optarg);
          if (amplification == 0) usage (argv[0]);

        case 'l':
          level_meters = 1;
          break;

        case 'i':
          recsrc = optarg;
          break;

        case 'm':
          nfiles = atoi (optarg);
          break;

        case 's':
          speed = atoi (optarg);
          if (speed == 0)
            {
              fprintf (stderr, "Bad sampling rate given\n");
              exit (-1);
            }
          if (speed < 1000) speed *= 1000;
          break;

        case 'r':
          c = snprintf (script, sizeof (script), "%s", optarg);
          if ((c >= sizeof (script)) || (c < 0))
            {
              fprintf (stderr, "-r argument is too long!\n");
              exit (-1);
            }
          break;

        case 't':
          datalimit = atoi (optarg);
          break;

        case 'w':
          break;

        case 'v':
          verbose = 1;
          break;

        case 'h':
        default:
          usage (argv[0]);
      }

  if (optind != argc - 1)	/* No file or multiple file names given */
    {
      usage (argv[0]);
      exit (-1);
    }

  if (format == -1) format = container_a[type].dformat;
  if (channels == -1) channels = container_a[type].dchannels;
  if (speed == -1) speed = container_a[type].dspeed;
  switch (format)
    {
      case AFMT_U8:
      case AFMT_S16_NE:
      case AFMT_S24_NE:
      case AFMT_S32_NE: break;
      default: level_meters = 0; /* Not implemented */
    }

  signal (SIGSEGV, fatal_signal);
  signal (SIGPIPE, fatal_signal);
  signal (SIGINT, intr);

  open_audio ();

  if (nfiles > 1)		/* Record multiple files */
    {
      int tmp_datalimit = datalimit;
      int i;

      for (i = 0; i < nfiles; i++)
	{
	  char tmpname[256];
	  int err;

	  sprintf (tmpname, argv[optind], i + 1);
	  datalimit = tmp_datalimit;
	  if ((err = do_record (dspdev, tmpname)) != 0)
	    exit (err);
	  fprintf (stderr, "\n");
	}
      if (*script)
	{
	  printf
	    ("Waiting for the '%s' script(s) to finish - please stand by\n",
	     script);
	  while (wait (NULL) != -1);
	}

      exit (0);
    }

  return do_record (dspdev, argv[optind]);
}
